File: leetcode/17.letter-combinations-of-a-phone-number/ts/lettercombinations.ts
const letterCombinations = (digits: string): string[] => {
    const letters = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
    const res: string[] = []
    const path: string[] = []

    const backtrack = (start: number) => {
        if (path.length === digits.length) {
            res.push(path.join(""))
            return
        }

        const options = letters[Number(digits[start]) - 2]

        for (let i = 0; i < options.length; i++) {
            path.push(options[i])
            backtrack(start + 1)
            path.pop()
        }
    }

    backtrack(0)

    return res
}

File: leetcode/17.letter-combinations-of-a-phone-number/go/lettercombinations.go
package lettercombinations

func letterCombinations(digits string) []string {
	letters := []string{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	res := []string{}
	path := []byte{}

	var backtrack func(int)

	backtrack = func(start int) {
		if len(path) == len(digits) {
			res = append(res, string(path))
			return
		}

		options := letters[int(digits[start]-'0')-2]
		for i := 0; i < len(options); i++ {
			path = append(path, options[i])
			backtrack(start+1)
			path = path[:len(path)-1]
		}
	}

	backtrack(0)
	return res
}
